### ISSUE TITLE
[reggen] Add tooling for checking in-line countermeasure annotations

### ISSUE THREAD

**Issue by msfschaffner** [lowRISC/opentitan#10071]
As discussed in the `entropy_src` D2S review meeting today, it would be great if we could annotate the code with the countermeasure IDs. 

These would not be exact annotations, but rather serve as hints where an auditor should look in order to audit a specific countermeasure. I.e., certain countermeasures can be pinpointed exactly  - think prim like instances like `prim_count`, whereas algorithmic countermeasures like masking schemes are a bit less well localized.

The tentative format for such annotations should be `// SEC_CM [INSTANCE].ASSET.CM_TYPE` in order to facilitate checks in reggen.

In particular, reggen should:
- Extract all lines starting with `// SEC_CM` from `ipname/rtl/*.sv` files
- Match them up against the Hjson list defined, and throw an error if
  - the countermeasure annotation does not exist in the Hjson list
  - a countermeasure from the Hjson list cannot be matched up with an RTL annotation (it is allowed to use the same countermeasure ID multiple times in RTL)

Possible future enhancements:
- Instead of just grepping files under `ipname/rtl/*.sv`, fusesoc could be invoked to build the file list
- In the documentation, we can append a list of links to each countermeasure ID that takes the reviewer/auditor to the code section with the annotation (on GitHub).

**Comment by msfschaffner** [comment]
CC @sriyerg

**Comment by tjaychen** [comment]
that sounds good.
so hypothetically speaking, let's say there's something called `CTRL.MUBI`, but in the design there are actually multiple instances of such.  

Each one would then probably match to the same entry in the hjson?

**Comment by msfschaffner** [comment]
That is correct, unless we want to enforce a 1:1 correspondence.
Note that such 1:1 correspondence would then probably have to be checked on a per SV module basis, since reggen cannot elaborate generate loops that create multiple instances of the same module.
Do we have any preference?

**Comment by tjaychen** [comment]
i dont think so. I think 1 to many should be fine, especially when it comes to the more standard types. 
If during the review process we find the need to split them up, then it might be up to the designer to actually tag them with separate prefixes so they can be separately identified.

**Comment by msfschaffner** [comment]
Yes, I think I am also leaning towards leaving more flexibility.
For the standardized ones, this would actually allow us to collapse them into one item, whereas for more custom stuff we probably want to call them out specifically.

This would eliminate quite some replication in blocks like OTP:
https://github.com/lowRISC/opentitan/blob/7df0097b04e67cda6797e3bfef3f9c7b99b638cf/hw/ip/otp_ctrl/data/otp_ctrl.hjson#L892-L909

**Comment by vogelpi** [comment]
I definitely favor flexibility, too. Collapsing multiple standardized CM instances seems more acceptable with the annotations in the RTL. However, there may still be cases if the list of CM for a particular instance are different in which case giving individual names simplifies things . For example, you may have different FSMs in an IP that all use the same base hardening but only one FSM features global escalation, too.

**Comment by mwbranstad** [comment]
Initial thought was that 1:1 would be more effective, but after more consideration, maybe just having the broad CM listed in the hjson and the instance name in the RTL is sufficient.

**Comment by msfschaffner** [comment]
@mwbranstad since we would like to globally refer to the same CM IDs, I think we should not use different IDs in RTL.

I.e., either the annotator chooses to make instance names explicit, and then the list in Hjson should reflect that, or they choose to use the general ID, in which case you get multiple tags of the same name in RTL.

**Comment by sriyerg** [comment]
+1 for flexibility - it will be useful to enforce for example, ALL instances of some critical counter measure `foo` are properly verified, once we start extracting testplan bits from these.

**Comment by cdgori** [comment]
As the reviewer, probably 1:1 would be easier to deal with.  But that basically makes the countermeasure lists unwieldy and forces a need for manual reconciliation which is not desirable.  So 1:many for things like MUBI is ok for me, I think.  We'll check the individual MUBIs in D2S and have to keep in mind that downstream changes to anything countermeasure-related (adding or removing) probably mean retriggering a D2S review.

**Comment by cdgori** [comment]
One other comment, what I am finding in the review process is that we are pursuing two goals in parallel:

1) for annotated/stated countermeasures, are they present / properly used?
2) are any countermeasures *missing* that should be present - this usually by enumerating assets and trying to think about attacks / potential design weaknesses or single points of failure.

The second one is harder without fully enumerated 1:1 type of lists.  I think the flexibility is still probably worth it, it just means that we spend a little more time on the review (a one-time cost, hopefully) to get everything accounted for.

**Comment by tjaychen** [comment]
btw i think one thing we "might" not be dealing with yet is the templated modules. I've run into a few cases where either the wrong hjson or maybe the wrong rtl was being inspected? do you happen to know?

**Comment by msfschaffner** [comment]
Ah, this is something we need to double check.

The regtool does the check on the files that live at `../rtl/` relative to the hjson, so it could be that this currently only works for generated IPs that are generated with IP gen, since that one copies the entire documentation over as well (and the docs are built from the generated folder)...

Could this be the issue?

**Comment by msfschaffner** [comment]
Dang, it looks like the checking currently does not work properly for generated IPs in general, no matter how they have been generated.

**Comment by msfschaffner** [comment]
Ok so #10271 should take care of IPs generated via IP gen. 
Still need to figure out the ones that are generated with the legacy mechanism.

**Comment by msfschaffner** [comment]
I pushed another patch to #10271 which takes care of legacy-style generated IPs, too.

**Comment by msfschaffner** [comment]
Ok generated IPs should now be covered.
However, we still need to figure out how we want to deal with vendored-in IP like Ibex, for instance.
We probably need a way to redirect reggen to include and scan vendored-in RTL files.

@rswarbrick @GregAC for thoughts.

**Comment by msfschaffner** [comment]
Testplan support for countermeasure items in Hjson is about to be added in #10556 by @sriyerg.

**Comment by tjaychen** [comment]
should we close this now? It seems like everything that's needed has been added.

**Comment by rswarbrick** [comment]
Were we keeping it open until we'd upgraded the warnings to errors?

**Comment by tjaychen** [comment]
ah that might be a while until i officially move all my stupid templated
blocks over to ip_autogen...

**Comment by msfschaffner** [comment]
@tjaychen I believe we should have support for both IPgen'd and legacy gen'd blocks... or does that not fully work on some of your blocks?

**Comment by tjaychen** [comment]
> @tjaychen I believe we should have support for both IPgen'd and legacy gen'd blocks... or does that not fully work on some of your blocks?

o it's not that. It's because some of my blocks are really dependent on top level (ie, the hjson alone isn't enough, it actually needs to wait for other blocks to be parsed), so the hw/ip/* stuff are not really up to date and will be missing certain statements, leading to warnings.

**Comment by msfschaffner** [comment]
ah ok, got it.

**Comment by msfschaffner** [comment]
One other mode that should be added: reggen cannot parse vendored code and prims at the moment, which means that we have to add all the SEC_CM labels in the comportable IP wrappers (see https://github.com/lowRISC/opentitan/pull/12104#discussion_r849935613). This could be solved, e.g., by leveraging the build system to gather the files...

**Comment by msfschaffner** [comment]
@tjshep we may want to fix and close this one as well when doing the IPgen migration.

**Comment by msfschaffner** [comment]
CC @matutem 

We should be able to complete this once everything works with Ipgen.

**Comment by matutem** [comment]
It turns out ipgen is not required: what needs to be done is to avoid running regtool in hw/Makefile for generated blocks, since topgen will run regtool with the right hjson file. This is addressed by https://github.com/lowRISC/opentitan/pull/21053.

There is also an issue (https://github.com/lowRISC/opentitan/issues/20887) to add missing countermeasure annotation to some blocks which issue warnings.

Once these two items are dealt with we could turn missing annotations into errors by default, perhaps with a flag to turn them into warnings for RTL development?

**Comment by vogelpi** [comment]
@matutem , I've assigned you to this issue because it is my understanding that you are in the process of finishing the last items before we can switch the warnings into errors. Please let me know if this is not correct, thanks!

**Comment by andreaskurth** [comment]
Matute has a PR that converts flash_ctrl to ipgen, and a follow-up PR will resolve this

**Comment by matutem** [comment]
An update: https://github.com/lowRISC/opentitan/pull/22848 has a change that triggers errors for any unimplemented countermeasures. Unfortunately, it is failing for pinmux because we invoke regtool.py against hw/top_earlgrey/ip/pinmux/data/autogen/pinmx.hjson, and that fails. The reason is that it cannot find all the sources, since they are split between hw/ip/pinmux and hw/top_earlgrey/ip/pinmux (for autogen code). It could be messy to implement a fix in regtool, and this problem won't affect pinmux if it was generated by ipgen.

I would like to commit this fix once pinmux is done via ipgen, which it should be done as part of #8440.

**Comment by andreaskurth** [comment]
Discussed in triage meeting, agreed to continue tracking this in M4
